sway-lsp/src/core/session.rs

```diff
@@ -80,166 +66,143 @@ impl Default for Session {
-                        })
+pub fn token_definition_response(
+    uri: &Url,
+    position: Position,
+    engines: &Engines,
+    token_map: &TokenMap,
+    sync: &SyncWorkspace,
+) -> Option<GotoDefinitionResponse> {
+    let _p = tracing::trace_span!("token_definition_response").entered();
+    token_map
+        .token_at_position(uri, position)
+        .and_then(|item| item.value().declared_token_ident(engines))
+        .and_then(|decl_ident| {
+            decl_ident.path.and_then(|path| {
+                // We use ok() here because we don't care about propagating the error from from_file_path
+                Url::from_file_path(path).ok().and_then(|url| {
+                    sync.to_workspace_url(url).map(|url| {
+                        GotoDefinitionResponse::Scalar(Location::new(url, decl_ident.range))
                     })
                 })
             })
-    }
-
-    pub fn completion_items(
-        &self,
-        uri: &Url,
-        position: Position,
-        trigger_char: &str,
-        token_map: &TokenMap,
-        engines: &Engines,
-    ) -> Option<Vec<CompletionItem>> {
-        let _p = tracing::trace_span!("completion_items").entered();
-        let shifted_position = Position {
-            line: position.line,
-            character: position.character - trigger_char.len() as u32 - 1,
-        };
-        let t = token_map.token_at_position(uri, shifted_position)?;
-        let ident_to_complete = t.key();
-        let fn_tokens = token_map.tokens_at_position(engines, uri, shifted_position, Some(true));
-        let fn_token = fn_tokens.first()?.value();
-        let compiled_program = &*self.compiled_program.read();
-        if let Some(TypedAstToken::TypedFunctionDeclaration(fn_decl)) = fn_token.as_typed() {
-            if let Some(program) = &compiled_program.typed {
-                return Some(capabilities::completion::to_completion_items(
-                    &program.namespace,
-                    engines,
-                    ident_to_complete,
-                    fn_decl,
-                    position,
-                ));
-            }
-        }
-        None
-    }
+        })
+}
 
-    /// Returns the [Namespace] from the compiled program if it exists.
-    pub fn namespace(&self) -> Option<Namespace> {
-        let compiled_program = &*self.compiled_program.read();
-        if let Some(program) = &compiled_program.typed {
-            return Some(program.namespace.clone());
-        }
-        None
+pub fn completion_items(
+    uri: &Url,
+    position: Position,
+    trigger_char: &str,
+    token_map: &TokenMap,
+    engines: &Engines,
+    compiled_programs: &CompiledPrograms,
+) -> Option<Vec<CompletionItem>> {
+    let _p = tracing::trace_span!("completion_items").entered();
+    let program = compiled_programs.program_from_uri(uri, engines)?;
+    let shifted_position = Position {
+        line: position.line,
+        character: position.character - trigger_char.len() as u32 - 1,
```

Comment:
nit: unrelated to PR here

`character: position.character - trigger_char.len() as u32 - 1,` may throw error (overflow) if `trigger_char` is empty (len = 1).
may want to update
```suggestion
        character: position.character.saturating_sub(trigger_char.len() as u32 + 1),
```